### PR TITLE
Integrate Docker

### PR DESCRIPTION
**Pull request by glenn-sorrentino**
Tested working locally

Created `Dockerfile` that follows our latest installer script as a template.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,53 @@
+FROM python:3.11-slim
+
+ARG DEBIAN_FRONTEND=noninteractive
+
+RUN apt-get update && apt-get -y dist-upgrade && \
+    apt-get install -y \
+    git \
+    nginx
```
Comment (brassy-endomorph) [inline Dockerfile:8]: We don't need `nginx`. The container will run a single process (`gunicorn`) and expose that. We'll use something like a docker-native reverse proxy (`traefik` most likely) to handle routing.

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,53 @@
+FROM python:3.11-slim
+
+ARG DEBIAN_FRONTEND=noninteractive
+
+RUN apt-get update && apt-get -y dist-upgrade && \
+    apt-get install -y \
+    git \
```
Comment (brassy-endomorph) [inline Dockerfile:7]: Why is `git` needed?

File: Dockerfile [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+    nginx
+
+RUN pip install virtualenv
+
+WORKDIR /frontpage
+
+COPY . /frontpage
+
+RUN rm -rf venv && virtualenv venv
```
Comment (brassy-endomorph) [inline Dockerfile:16]: Don't use a `venv` in the container. Just install the packages directly.

File: Dockerfile [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+RUN . venv/bin/activate && \
+    pip install -U pip && \
+    pip install -U poetry && \
+    poetry self add poetry-plugin-export && \
+    export PYTHON_KEYRING_BACKEND=keyring.backends.null.Keyring && \
+    poetry export --only main -f requirements.txt --output requirements.txt && \
```
Comment (brassy-endomorph) [inline Dockerfile:23]: You should look up the [builder pattern](https://docs.docker.com/build/building/multi-stage/). First stage just install `poetry` and export the requirements. Second stage copies the generated `requirements.txt` the use that to install, and do so directly with `pip` and not `poetry`. Poetry shouldn't be needed at all in the final container.

File: Dockerfile [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+    pip install -r requirements.txt && \
+    pip install gunicorn
+
+RUN touch .env && \
```
Comment (brassy-endomorph) [inline Dockerfile:27]: Containers shouldn't create state like this. The `.env` file can be mounted as a read-only volume, but more ideally you'd launch the container with the `env_file` configuration.

File: Dockerfile [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+        echo "FLASK_SECRET_KEY=$FLASK_SECRET_KEY" >> .env; \
+    fi
+
+ENV FLASK_APP=frontpage:app
+
+RUN . venv/bin/activate && \
+    poetry run flask db upgrade && \
+    poetry run flask db-extras add-article-types
+
+RUN chown -R www-data:www-data /frontpage && \
```
Comment (brassy-endomorph) [inline Dockerfile:40]: You shouldn't create directories like this. These directories are going to be mounted volumes when the container runs, and all their properties should be set external to the container.

File: Dockerfile [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+    mkdir -p /frontpage/frontpage/static/uploads && \
+    chown -R www-data:www-data /frontpage/frontpage/static/uploads && \
+    chmod -R 755 /frontpage/frontpage/static/uploads
+
+RUN apt-get install nginx -y && \
```
Comment (brassy-endomorph) [inline Dockerfile:45]: Second `nginx` install?

File: Dockerfile [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+    cp files/nginx.conf /etc/nginx/sites-available/frontpage && \
+    ln -sf /etc/nginx/sites-available/frontpage /etc/nginx/sites-enabled && \
+    rm -f /etc/nginx/sites-available/default && \
+    rm -f /etc/nginx/sites-enabled/default
+
+EXPOSE 5000
+
+CMD ["sh", "-c", ". venv/bin/activate && gunicorn --bind 0.0.0.0:5000 frontpage:app & nginx -g 'daemon off;'"]
```
Comment (brassy-endomorph) [inline Dockerfile:53]: Running `sh -c 'command_1 & command_2'` is bad practice. If the `gunicorn` process crashes, we'll never know. Containers should only ever have one process, and when that "goes wrong" it should exit and the orchestrator (docker swarm, kubernetes) should handle restarting or whatever.

**Review by brassy-endomorph — CHANGES_REQUESTED** [review]

File: Dockerfile (hunk 2 above)
Comment (glenn-sorrentino) [reply Dockerfile:7]: Left over from old code.

File: Dockerfile (hunk 3 above)
Comment (glenn-sorrentino) [reply Dockerfile:16]: 👍

File: Dockerfile (hunk 4 above)
Comment (glenn-sorrentino) [reply Dockerfile:23]: 👍

File: Dockerfile (hunk 7 above)
Comment (glenn-sorrentino) [reply Dockerfile:45]: 👍

File: Dockerfile (hunk 8 above)
Comment (glenn-sorrentino) [reply Dockerfile]: 👍

File: Dockerfile (hunk 5 above)
Comment (glenn-sorrentino) [reply Dockerfile]: 👍

File: Dockerfile (hunk 6 above)
Comment (glenn-sorrentino) [reply Dockerfile]: 👍

File: Dockerfile (hunk 8 above)
Comment (glenn-sorrentino) [reply Dockerfile]: 👍

**Pull request closed without merging**